### PR TITLE
[WIP][SPARK-37877][SQL] Support clear shuffle dependencies eagerly for thrift server

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In this PR, we add a config to control whether thrift server operations will clear shuffle dependencies eagerly after being executed.

In long-running applications, like thrift server, with large driver JVMs, where there is little memory pressure on the driver, the driver gc may happen very occasionally or not at all. Not cleaning at all may lead to executors running out of disk space after a while.

### Why are the changes needed?

For the thrift server, it currently relies on a periodical system.gc to release shuffle meta/data e.t.c, which is not efficient enough towards its usage scenario as the gc does not always occur immediately. Unclean data may cause thrift server memory issues, like OOM, or disk issues on the work side, like no space left for the device.

### Does this PR introduce _any_ user-facing change?

added new conf

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
FWIW, there's a bug on updating the status in the check (https://github.com/apache/spark/pull/35179#issuecomment-1011881076). We should check the actual tests e.g., at https://github.com/yaooqinn/spark/runs/4798556788

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 1]
```diff
@@ -332,6 +339,10 @@ private[hive] class SparkExecuteStatementOperation(
           HiveThriftServer2.eventManager.onStatementFinish(statementId)
         }
       }
+      if (cleanShuffleDeps && rdd != null) {
+        rdd.cleanShuffleDependencies()
+        rdd = null
```
Comment (mridulm) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:344]: When `THRIFTSERVER_INCREMENTAL_COLLECT` is `true`, we will end up cleaning up dependencies before we have used up the data from df.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 2]
```diff
@@ -288,10 +293,12 @@ private[hive] class SparkExecuteStatementOperation(
...
 
       sqlContext.sparkContext.setJobGroup(statementId, substitutorStatement, forceCancel)
-      result = sqlContext.sql(statement)
-      logDebug(result.queryExecution.toString())
+      val df = sqlContext.sql(statement)
+      logDebug(df.queryExecution.toString())
       HiveThriftServer2.eventManager.onStatementParsed(statementId,
-        result.queryExecution.toString())
+        df.queryExecution.toString())
+      rdd = df.rdd
```
Comment (mridulm) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:300]: `df.rdd` has tripped me in the past - the fact that it actually ends up executing the prefix dag was surprising to me !

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:344]: thanks for the input

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala:300]: the changes here cannot trigger the eager clean for all cases, for example, CTAS, where the `S` contains a shuffle, will not be able to be cleared. The possible and proper place may be SparkPlan

**Comment by yaooqinn** [comment]
Still in POC, any inputs are welcome :)

**Pull request closed without merging**